### ISSUE TITLE
Revise the plugin discovery mechanism for converters

### ISSUE THREAD

**Issue by blagae** [opensagres/xdocreport#73]
Currently, the plugin discovery process asks the `IConverterDiscovery` implementation for its `getFrom()`, `getTo()`, and `getVia()` methods. This is a poignant problem for me personally now that there are two iText implementations which both do DOCX to PDF conversion via XWPF. In general, it can get in the way of anyone who is trying to develop a new converter.

The `IConverter::isDefault()` would be of limited use, because I can't imagine anyone writing a converter and then not wanting it to be used if there's an alternative :-) 

Do you agree that this should be reviewed ? I am willing to cooperate for a implementation.

**Comment by angelozerr** [comment]
@blagae perhaps we should have XWPF_ITEXT and XWPF_ITEXT5 inseatd of having XWP for https://github.com/opensagres/xdocreport/blob/master/converter/fr.opensagres.xdocreport.converter/src/main/java/fr/opensagres/xdocreport/converter/ConverterTypeVia.java ?

Otherwise if you have a better idea, don't hesitate to share it.

**Comment by blagae** [comment]
I was thinking about this, and I concluded that your suggestion would limit extendability for custom-made plugins. Any new plugin would have to either fork and add an enum value and build from source, or create a PR to this project to get registered 'officially' in the enum. In the end, I thought it would be better to add an extra method to the Options class, something like ...

    public Options with(String pluginId)

... which could be chained after `getTo()` and/or `via()`.

So usage might be:

    Options options = Options.getTo(ConverterTypeTo.PDF)
                             .via(ConverterTypeVia.XWPF) // optional ?
                             .with("Docx2PDFViaIText");

"Docx2PDFViaIText" is the current return value of `XWPF2PDFViaITextConverterDiscovery::getId()`

In the discovery mechanism, we'd have to add another layer of logic if there are multiple implementations with the same from, to, and via. It would query the `IConverterDiscovery::getId()` method for equality with the input String, if any.

**Comment by blagae** [comment]
I have made a candidate implementation, and started writing tests. https://github.com/blagae/xdocreport/tree/feature/optionswith

There are more tests to come, and you're welcome to add some others for yourself. Since it's unverified functionality for now, it shouldn't be pulled into master yet.

**Comment by angelozerr** [comment]
@blagae its' a very good idea. I like it! I think we should add too a new enum ConverterTypeWith with ITEXT2, ITEXT5 liek https://github.com/opensagres/xdocreport/blob/master/converter/fr.opensagres.xdocreport.converter/src/main/java/fr/opensagres/xdocreport/converter/ConverterTypeVia.java What do you think?

**Comment by blagae** [comment]
If you want that, it's fine by me, as long as users can write a String of their own choice, too (as for the other settings). Do you want to add a `getWith()` method to `IConverterDiscovery`, or is `getId()` good enough ? A new method would break the stability of the API, of course, but it might be cleaner than using an existing method.

**Comment by angelozerr** [comment]
> If you want that, it's fine by me, as long as users can write a String of their own choice, too (as for the other settings)

Yes we have designed XDocReport like this : provides enum but can use String if needed.

> A new method would break the stability of the API, of course, but it might be cleaner than using an existing method.

No problem with "break the stability of the API". I think only XDocReport uses this API.

**Comment by blagae** [comment]
> Yes we have designed XDocReport like this : provides enum but can use String if needed.

OK that sounds good, I'll add it to my branch.

> No problem with "break the stability of the API". I think only XDocReport uses this API.

Any users with a custom discovery project would be impacted. But of course, breaking their code might be a good way to let them know about this new method and possibility :-)

**Comment by blagae** [comment]
Ok I've got a pull request ready, so I'll close this. But first some more questions to @angelozerr :

* I'd like to create a discovery project for iText 5 as well. I already made a test run of this, and I noticed that I'll need to refactor the package name of the fr.opensagres.xdocreport.itext5.extension and some other stuff. Is that ok with you ?
* ODF to PDF apparently doesn't use the discovery mechanism right now. Do you want this to be implemented ?

**Comment by angelozerr** [comment]
> Is that ok with you ?

If you don't rename some package user API, there ar eno problem with that.

> ODF to PDF apparently doesn't use the discovery mechanism right now.

Are you sure with that?

https://github.com/opensagres/xdocreport/blob/master/converter/fr.opensagres.xdocreport.converter.odt.odfdom/src/main/resources/META-INF/services/fr.opensagres.xdocreport.converter.discovery.IConverterDiscovery

**Comment by blagae** [comment]
I'm not sure, but it's the best way that I can currently explain that my local ODF tests don't need the discovery module to work. At least, when I debug, I never enter discovery logic for ODF, and I do for DOCX.

In the convert method of `AbstractXDocReport`, you see that the discovery logic is called:
https://github.com/opensagres/xdocreport/blob/master/document/fr.opensagres.xdocreport.document/src/main/java/fr/opensagres/xdocreport/document/AbstractXDocReport.java#L690

In the convert method of ODF's `PdfConverter`, this doesn't happen. https://github.com/opensagres/xdocreport/blob/master/thirdparties-extension/org.odftoolkit.odfdom.converter.pdf/src/main/java/org/odftoolkit/odfdom/converter/pdf/PdfConverter.java

To further make my point: the `Options` class for ODF has a completely different structure and behavior than the one for DOCX:
https://github.com/opensagres/xdocreport/blob/master/thirdparties-extension/org.odftoolkit.odfdom.converter.core/src/main/java/org/odftoolkit/odfdom/converter/core/Options.java

I'm not saying that this has to change. But if you want it to, then I can help.

**Comment by angelozerr** [comment]
At first you must understand that you have :

 * converters implementation stored in https://github.com/opensagres/xdocreport/tree/master/thirdparties-extension. Those implementation doesn't use discovery logic because there are used just to convert a document to another document with a format.

 * reporting with convertion. It means that you generate a docx report from a docx template and you wish to convert it. Here discovery logic is used.

> In the convert method of AbstractXDocReport, you see that the discovery logic is called:
https://github.com/opensagres/xdocreport/blob/master/document/fr.opensagres.xdocreport.document/src/main/java/fr/opensagres/xdocreport/document/AbstractXDocReport.java#L690

Yes it's for generate a report and convert it to another format (not for just convert a document to another format)

> In the convert method of ODF's PdfConverter, this doesn't happen. https://github.com/opensagres/xdocreport/blob/master/thirdparties-extension/org.odftoolkit.odfdom.converter.pdf/src/main/java/org/odftoolkit/odfdom/converter/pdf/PdfConverter.java

Yes because it's just to convert document to another format. You have the same thing for docx https://github.com/opensagres/xdocreport/blob/master/thirdparties-extension/org.apache.poi.xwpf.converter.pdf/src/main/java/org/apache/poi/xwpf/converter/pdf/PdfConverter.java It doesnt use discovery logic too.

> To further make my point: the Options class for ODF has a completely different structure and behavior than the one for DOCX:
https://github.com/opensagres/xdocreport/blob/master/thirdparties-extension/org.odftoolkit.odfdom.converter.core/src/main/java/org/odftoolkit/odfdom/converter/core/Options.java

You hav ethe same thing for docx https://github.com/opensagres/xdocreport/blob/master/thirdparties-extension/org.apache.poi.xwpf.converter.core/src/main/java/org/apache/poi/xwpf/converter/core/Options.java

If you se the code it's the same code than ODT, but we decide to do that to avoid having a new commons project for converter.

**Comment by blagae** [comment]
So that means I was creating wrong client code for ODT. Good to know :-)
